core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -173,6 +174,24 @@ private[scheduler] class BlacklistTracker (
         listenerBus.post(SparkListenerExecutorBlacklisted(now, exec, newTotal))
         executorIdToFailureList.remove(exec)
         updateNextExpiryTime()
+        // Add executor to blacklist before attempting to kill it. This allows a scheduler backend
+        // to immediately fail to allocate resources on this executor, since killing could be
+        // asynchronous.
```

Comment:
I put this comment here to avoid a race between an individual executor being blacklisted, and being scheduled for killing. The comment makes more sense in the context of an individual executor, and not an entire node.

What you're talking about is something different, and more complex: the need to - explicitly - blacklist all the executors on a node before marking that node for killing.

For example, consider hostA with executors 1, 2, and 3. We can blacklist 1, and then 2, and then hostA, but 3 is never explicitly blacklisted. The `nodeBlacklist` case needs to blacklist 3, which means we need some explicit mapping between hosts and the executors on that host in BlacklistTracker. To date we've been trying to avoid exactly this to "keep the BlacklistTracker simple".